### PR TITLE
[ML] Add the number of queue requests per node to deployment stats

### PR DESCRIPTION
**Pull request by davidkyle**
Every node hosting a model has a queue of inference requests, the number of items in that queue is a useful indicator of throughput & performance. The size is now reported in the deployment node stats in the new field `number_of_requests_in_queue`.

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/deployment/DeploymentManager.java [hunk 1]
```diff
@@ -103,7 +103,8 @@ public void startDeployment(TrainedModelDeploymentTask task, ActionListener<Trai
             .map(
                 processContext -> new ModelStats(
                     processContext.getResultProcessor().getTimingStats(),
-                    processContext.getResultProcessor().getLastUsed()
+                    processContext.getResultProcessor().getLastUsed(),
+                    processContext.executorService.queueSize()
```
[the remark this answers is not in the document]
Comment (davidkyle) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/deployment/DeploymentManager.java]: 👍  `+ pendingResults.size()` is a good idea

**Comment by dimitris-athanasiou** [comment]
I think we need to be careful here about the naming of this. I think it's misleading to include pending results in `number_of_requests_in_queue`. I think the queue size is useful for a user trying to determine the right value for the queue capacity. This shouldn't include the requests that have made it out of the queue and are now in the process buffer. I also think the total number of pending results would be useful, where that would be the queue size plus the pending results (as we have them in code). For the user, this would be `total_pending_requests`, as what we call `pending_results` in code doesn't make sense for the user.

Thus I suggest we have two different values, one for the queue size and another for the total pending requests count. What do you think?

**Comment by benwtrent** [comment]
@dimitris-athanasiou 

🤔 Yeah, its tricky. I agree `number_of_requests_in_queue` implies a relation to the configurable queue size. Including the pending/queued native requests could inflate that number and be confusing.

`total_pending_requests` is a good field name. 

I wonder if we need the queue size separate from the total pending requests? Increasing queue size increases the number of pending requests we allow...

**Comment by dimitris-athanasiou** [comment]
Thinking about it, even if we did report queue size, it still does not help the user know what value to set the queue capacity in the case of 429s. The only benefit of having queue size is that a user could detect the number getting too high and increase the queue capacity proactively. Not sure this is necessary, at least at this stage.

So yeah, maybe we can stick to a single number here and call it `number_of_pending_requests` or something.

**Comment by davidkyle** [comment]
A single figure is the right choice to avoid too many confusing stats and `number_of_pending_requests` is a good name.

> Thinking about it, even if we did report queue size, it still does not help the user know what value to set the queue capacity in the case of 429s. The only benefit of having queue size is that a user could detect the number getting too high and increase the queue capacity proactively. Not sure this is necessary, at least at this stage.

For a cluster with a single ML node and single queue you can infer all you need about the number of items in the queue from the presence of 429s but for multiple ML nodes it answers questions such as "is there a hot node" or "which nodes are overloaded". If you have peaky usage then increasing the queue size to smooth input will help with 429s but if the usage is sustained the solution is to add another ML node and deploy the model there. Monitoring queue size would help with that decision if you do not know the ingest rates. 

If those scenarios sound too hypothetical it's a valid concern, I'm happy to park this PR for now while we learn more about debugging deployments. 

As an aside this makes me thing we should include the node name in error messages in the case of timeouts and/or too many requests.

**Comment by davidkyle** [comment]
The inference routing code has this TODO

https://github.com/elastic/elasticsearch/blob/12ad399c488f0cc60e19b5e1b29c6d569cb4351a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportInferTrainedModelDeploymentAction.java#L79

A service that periodically monitored the deployment queue sizes on the ML nodes could make informed decisions about where to route a request.

**Comment by dimitris-athanasiou** [comment]
I think having `number_of_pending_requests` is an improvement so I'd rather we add it in now and improve in the future.

**Review by benwtrent — APPROVED** [review]

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM2

**Pull request merged**